src/grouping-and-mapping/clients/GroupsClient.ts

```diff
@@ -0,0 +1,107 @@
+
+  public async createGroup(accessToken: string, mappingId: string, group: GroupCreate): Promise<Group> {
+    if (!this.isSimpleIdentifier(group.groupName)) {
+      throw new RequiredError(
+        "groupName",
+        "Field groupName was invalid.",
+      );
+    }
+
+    if(this.isNullOrWhitespace(group.query)) {
+      throw new RequiredError(
+        "query",
+        "Required field query was null or undefined.",
+      );
+    }
+
+    const url = `${this._groupsUrl}/${encodeURIComponent(mappingId)}/groups`;
+    const requestOptions: RequestInit = this.createRequest("POST", accessToken, JSON.stringify(group));
+    return (await this.fetchJSON<GroupContainer>(url, requestOptions)).group;
+  }
+
+  public async deleteGroup(accessToken: string, mappingId: string, groupId: string): Promise<Response> {
+    const url = `${this._groupsUrl}/${encodeURIComponent(mappingId)}/groups/${encodeURIComponent(groupId)}`;
+    const requestOptions: RequestInit = this.createRequest("DELETE", accessToken);
+    return this.fetchJSON<Response>(url, requestOptions);
+  }
+
+  public async updateGroup(accessToken: string, mappingId: string, groupId: string, group: GroupUpdate): Promise<Group> {
+    if(null == group.groupName && null == group.description && null == group.query) {
+      throw new RequiredError(
+        "group",
+        "All properties of group were missing.",
+      );
+    }
+
+    if (null != group.groupName && !this.isSimpleIdentifier(group.groupName)) {
+      throw new RequiredError(
+        "groupName",
+        "Field groupName was invalid.",
+      );
+    }
+
+    if (null != group.query && this.isNullOrWhitespace(group.query)) {
+      throw new RequiredError(
+        "query",
+        "Required field query was null or undefined.",
+      );
+    }
+
+    const url = `${this._groupsUrl}/${encodeURIComponent(mappingId)}/groups/${encodeURIComponent(groupId)}`;
+    const requestOptions: RequestInit = this.createRequest("PATCH", accessToken, JSON.stringify(group));
+    return (await this.fetchJSON<GroupContainer>(url, requestOptions)).group;
+  }
+
+  public async getGroup(accessToken: string, mappingId: string, groupId: string): Promise<Group> {
+    const url = `${this._groupsUrl}/${encodeURIComponent(mappingId)}/groups/${encodeURIComponent(groupId)}`;
+    const requestOptions: RequestInit = this.createRequest("GET", accessToken);
+    return (await this.fetchJSON<GroupContainer>(url, requestOptions)).group;
+  }
+
+  public async getGroups(accessToken: string,  mappingId: string, top?: number | undefined): Promise<GroupList> {
+    if(!this.topIsValid(top)) {
+      throw new RequiredError(
+        "top",
+        "Parameter top was outside of the valid range [1-1000]."
+      );
+    }
+
+    const url = top ? `${this._groupsUrl}/${encodeURIComponent(mappingId)}/groups?$top=${top}` : `${this._groupsUrl}/${encodeURIComponent(mappingId)}/groups`;
+    const request = this.createRequest("GET", accessToken);
```

Comment:
```
const baseUrl = `${this._groupsUrl}/${encodeURIComponent(mappingId)}/groups`;
const url = `${baseUrl}${top ? `?$top=${top}` : ''}`;
```

Reply:
I opted for something like this instead: 
`private _baseUrl = `${this.groupingAndMappingBasePath}/datasources/imodel-mappings`;`

` protected constructUrl(mappingId: string, groupId?: string, top?: number): string {
    if(groupId)
      return `${this._baseUrl}/${encodeURIComponent(mappingId)}/groups/${encodeURIComponent(groupId)}`;

    if(top)
      return `${this._baseUrl}/${encodeURIComponent(mappingId)}/groups?$top=${top}`;

    return `${this._baseUrl}/${encodeURIComponent(mappingId)}/groups`;
  }
`
Since properties will also have a helper function to construct the endpoint. What do you think?

Reply:
Opted for helper function to construct the endpoint, since properties will have one too to keep consistent and for readability.